File: app/src/main/java/com/left/gank/ui/discovered/DiscoveredFragment.java
package com.left.gank.ui.discovered;

import android.content.res.Resources;
import android.os.Bundle;
import android.util.TypedValue;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;
import com.left.gank.R;
import com.left.gank.ui.base.LazyFragment;
import com.left.gank.ui.base.fragment.ButterKnifeFragment;
import com.left.gank.ui.discovered.jiandan.JiandanFragment;
import com.left.gank.ui.discovered.more.DiscoveredAdapter;
import com.left.gank.ui.discovered.more.DiscoveredMoreFragment;
import com.left.gank.ui.discovered.teamBlog.TeamBlogFragment;
import com.left.gank.ui.discovered.technology.TechnologyFragment;
import com.left.gank.ui.discovered.video.VideoFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * 发现
 * Create by LingYan on 2016-07-01
 */
public class DiscoveredFragment extends ButterKnifeFragment implements ViewPager.OnPageChangeListener {
    private static final String TYPE_VIDEO = "视频";
    private static final String TYPE_JIANDAN = "新鲜事";
    private static final String TYPE_THCHNOLOGY = "科技资讯";
    private static final String TYPE_TEAM_BLOG = "团队博客";
    private static final String TYPE_MORE = "更多";

    @BindView(R.id.discovered_tabLayout)
    TabLayout mTabLayout;
    @BindView(R.id.discovered_view_pager)
    ViewPager mViewPager;

    private List<String> mTitles;

    @Override
    protected int fragmentLayoutId() {
        return R.layout.fragment_discovered;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        List<LazyFragment> mList = new ArrayList<>();
        mList.add(new VideoFragment());
        mList.add(new JiandanFragment());
        mList.add(new TechnologyFragment());
        mList.add(new TeamBlogFragment());
        mList.add(new DiscoveredMoreFragment());

        mTitles = new ArrayList<>();
        mTitles.add(TYPE_VIDEO);
        mTitles.add(TYPE_JIANDAN);
        mTitles.add(TYPE_THCHNOLOGY);
        mTitles.add(TYPE_TEAM_BLOG);
        mTitles.add(TYPE_MORE);

        DiscoveredAdapter mPagerAdapter = new DiscoveredAdapter(getChildFragmentManager(), mList,
                mTitles);
        mViewPager.setAdapter(mPagerAdapter);
        mViewPager.setOffscreenPageLimit(mPagerAdapter.getCount());
        mViewPager.addOnPageChangeListener(this);

        initTabLayout();
    }

    private void initTabLayout() {
        for (int i = 0; i < mTitles.size(); i++) {
            mTabLayout.addTab(mTabLayout.newTab().setText(mTitles.get(i)));
        }

        mTabLayout.setupWithViewPager(mViewPager);
        mTabLayout.setTabMode(TabLayout.MODE_FIXED);
        mTabLayout.setSelectedTabIndicatorColor(getContext().getResources().getColor(R.color.white));
    }

    private void refreshUi() {
        TypedValue typedValue = new TypedValue();
        Resources.Theme theme = getActivity().getTheme();
        theme.resolveAttribute(R.attr.colorPrimary, typedValue, true);
        int background = typedValue.data;
        mTabLayout.setBackgroundColor(background);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
    }

    @Override
    public void onPageScrollStateChanged(int state) {
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }
}

File: app/src/main/java/com/left/gank/ui/discovered/DiscoveredFragment.kt
package com.left.gank.ui.discovered

import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import com.left.gank.R
import com.left.gank.ui.base.fragment.ButterKnifeFragment
import com.left.gank.ui.discovered.jiandan.JiandanFragment
import com.left.gank.ui.discovered.more.DiscoveredAdapter
import com.left.gank.ui.discovered.more.DiscoveredMoreFragment
import com.left.gank.ui.discovered.teamBlog.TeamBlogFragment
import com.left.gank.ui.discovered.technology.TechnologyFragment
import com.left.gank.ui.discovered.video.VideoFragment
import kotlinx.android.synthetic.main.fragment_discovered.*

/**
 * 发现
 * Create by LingYan on 2016-07-01
 */
class DiscoveredFragment : ButterKnifeFragment() {
    private val titles: List<String> = listOf(
            TYPE_VIDEO,
            TYPE_JIAN_DAN,
            TYPE_THCHNOLOGY,
            TYPE_TEAM_BLOG,
            TYPE_MORE
    )

    private val fragments = listOf(
            VideoFragment(),
            JiandanFragment(),
            TechnologyFragment(),
            TeamBlogFragment(),
            DiscoveredMoreFragment()
    )

    override fun fragmentLayoutId(): Int = R.layout.fragment_discovered

    private val pageChangeListener = object : ViewPager.OnPageChangeListener {
        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

        }

        override fun onPageSelected(position: Int) {

        }

        override fun onPageScrollStateChanged(state: Int) {

        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        discovered_view_pager.apply {
            val mPagerAdapter = DiscoveredAdapter(childFragmentManager, fragments, titles)
            adapter = mPagerAdapter
            offscreenPageLimit = mPagerAdapter.count
            addOnPageChangeListener(pageChangeListener)
        }

        initTabLayout()
    }

    private fun initTabLayout() {
        discovered_tabLayout.apply {
            for (i in titles.indices) {
                addTab(newTab().setText(titles[i]))
            }
        }.also {
            it.setupWithViewPager(discovered_view_pager)
            it.tabMode = TabLayout.MODE_FIXED
            it.setSelectedTabIndicatorColor(ContextCompat.getColor(context!!, R.color.white))
        }
    }

    companion object {
        private const val TYPE_VIDEO = "视频"
        private const val TYPE_JIAN_DAN = "新鲜事"
        private const val TYPE_THCHNOLOGY = "科技资讯"
        private const val TYPE_TEAM_BLOG = "团队博客"
        private const val TYPE_MORE = "更多"
    }
}